### PR TITLE
REF: tighten signature on maybe_convert_platform

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx
- [ ] tests added / passed
- [ ] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/core/arrays/interval.py [hunk 1]
```diff
@@ -1616,7 +1613,15 @@ def _maybe_convert_platform_interval(values) -> ArrayLike:
...
         # prohibited for IntervalArray, so coerce to integer instead
         return np.array([], dtype=np.int64)
+    elif not is_list_like(values) or isinstance(values, ABCDataFrame):
+        # This will raise later, but we avoid passing to maybe_convert_platform
+        return values
     elif is_categorical_dtype(values):
         values = np.asarray(values)
+    elif not hasattr(values, "dtype") and not isinstance(values, (list, tuple, range)):
+        # TODO: should we just cast these to list?
+        return values
```
Comment (jreback) [inline pandas/core/arrays/interval.py:1623]: yes?

File: pandas/core/dtypes/cast.py [hunk 2]
```diff
@@ -119,16 +119,21 @@
...
     """ try to do platform conversion, allow ndarray or list here """
     if isinstance(values, (list, tuple, range)):
-        values = construct_1d_object_array_from_listlike(values)
-    if getattr(values, "dtype", None) == np.object_:
-        if hasattr(values, "_values"):
-            values = values._values
-        values = lib.maybe_convert_objects(values)
+        arr = construct_1d_object_array_from_listlike(values)
+    else:
+        # The caller is responsible for ensuring that we have np.ndarray
```
Comment (jreback) [inline pandas/core/dtypes/cast.py:129]: we should tighten this and return a np.ndarray

Comment (jbrockmendel) [reply pandas/core/dtypes/cast.py:129]: im planning to continue tightening, yes.  But im also hoping to start passing `convert_datetime=True, convert_timedelta=True` to lib.maybe_convert_objects on L134.

File: pandas/core/arrays/interval.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/arrays/interval.py:1623]: lets save for a separate pass, as that will be an API change

File: pandas/core/internals/construction.py [hunk 3]
```diff
@@ -400,7 +400,15 @@ def _prep_ndarray(values, copy: bool = True) -> np.ndarray:
             return arr[..., np.newaxis]
 
         def convert(v):
-            return maybe_convert_platform(v)
+            if not is_list_like(v) or isinstance(v, ABCDataFrame):
```
Comment (jreback) [inline pandas/core/internals/construction.py:403]: this seems like a pretty general function (just commenting nothing to do now)

Comment (jbrockmendel) [reply pandas/core/internals/construction.py:403]: yah im hoping we'll end up with just one sanitize_array-like function by the end of this process.  the hard part at this point is deciding what we want to do about #40114, #40110, and #24435

**Pull request merged**